feat(report): statistics in code coverage report

galtm opened the pull request:
This pull request adds summary data to XSLT code coverage reports, making the report more useful as well as more consistent with the test result reports.

@birdya22 did most of the design and implementation via #2067 -- thanks!! I reviewed the design and wrote some additional code, mostly on the testing side.

### Key features

- "XSpec:" and "Tested:" lines at top. Test result reports already have this information.
- **Contents** table with a row for each XSLT module. Columns indicate whether the module is used; # of hit nodes; # of missed nodes; # of "unknown coverage" nodes; and # of lines in that module. The top row of the table shows totals for the numeric columns. The path of the module is linked to the report section about that module.
- Header for each module shows the numbers of hit, missed, and unknown-coverage nodes. It no longer shows the number of lines in that module, but you can see that at the top.

#### Future enhancements

- Version numbers for Saxon and XSpec, because we know the versions can affect coverage reports. This addition will be part of #1404.
- Possibly more statistics

@birdya22, if I left out something that should be included here, please add a comment or edit this description.

### Zoomed-out view of an entire report
![image](https://github.com/user-attachments/assets/8c9e1a17-bb37-45f7-a8fe-6ff2d0b72157)

### Readable view of top of report
![image](https://github.com/user-attachments/assets/63181d8d-7cb3-470d-864e-1e28eaa81b0a)

### Interactive testing
I viewed many of the expected-result coverage reports under test/end-to-end/cases-coverage after regenerating them. I varied the browser width, clicked links, and changed the font size.

I also generated some coverage reports where the modules have long file paths. The only difference I see between long file paths and long scenario labels in the test result report is that a scenario label is generally a human-language description that includes spaces where the browser can wrap to the next line, whereas a file path might not have characters where the browser will wrap. So, you might end up with a horizontal scroll bar.

---
Fixes #2067.

galtm commented on test/end-to-end/processor/coverage/_normalizer.xsl:
```diff
@@ -47,17 +47,18 @@
 
 	<!--
-		Normalizes the links to the tested module
+		Normalizes the links to the test, the tested module, and sections of this report
```
FYI in case you want to know what this file is for: the test cases under `test/end-to-end` compare expected reports with newly generated reports, and part of the process is to normalize both reports in the pair prior to comparison. This stylesheet's job is to identify what fragments need normalization and what to replace them with. For instance, this stylesheet is why the reports under `test/end-to-end/cases-coverage/expected/stylesheet` have headers like `module: xsl-analyze-string-01.xsl` with only a filename instead of a full file path.

galtm commented on src/reporter/coverage-report.xsl:
```diff
@@ -46,6 +46,35 @@
+      select="replace(base-uri(), '-coverage\.xml', '-result.xml')" />
+   <xsl:variable name="result-doc" as="document-node(element(x:report))">
+      <xsl:choose>
+         <xsl:when test="doc-available($result-uri)">
+            <xsl:sequence select="doc($result-uri)"/>
+         </xsl:when>
+         <xsl:otherwise>
+            <!-- End-to-end coverage tests don't have access to *-result.xml.
```
I believe a standard end-user situation would reach `xsl:when` because the test result report is created immediately before the coverage report. Even if the user has customized the paths to the HTML reports, the two XML files should be consistent with the definition of `result-uri`. However, this repo's end-to-end testing does some things differently, so I had to add the `xsl:otherwise` element.

birdya22 commented on src/reporter/coverage-report.xsl:
```diff
@@ -70,6 +99,35 @@
+         <module uri="{$stylesheet-uri}" number-of-lines="{$number-of-lines}"
+            formatted-uri="{fmt:format-uri($stylesheet-uri)}" module-id="{$module-id}" xmlns="">
+            <xsl:if test="not(empty($module-id))">
+               <xsl:for-each select="doc($stylesheet-uri)/descendant::node()">
```
Indentation is going to be an issue, as it probably occurs on all lines except 3 (the xml PI and start/end `xsl:stylesheet` tags), so it might add up to nearly 50% of all `<coverage>` elements.  I'm not so worried about comments. So I think a predicate expression to eliminate whitespace only text nodes is a good idea. 
As an alternative to excluding indenting whitespace only nodes, is it worth considering removing the class attribute, or even the span element for indentation. I realise this isn't as easy as it sounds?

galtm replied:
Thanks for your input. 71e085212bd712cb8d30d41e3036c711a842503d excludes insignificant text nodes from the iteration. I ended up using `except` instead of a predicate expression because I thought it was easier to read and understand than double `not`s.

birdya22 commented:
> @birdya22 : Do you have an opinion about whether your TestStylesheet01.xspec file and its related XSLT files should be included in this repo?

Yes, as I think it is the only test that includes both 'used' and 'not used' modules. It also includes all types of coverage. However, the  file names I used aren't useful, so I have amended them and attached the new set.

[coverage-statistics-test.zip](https://github.com/user-attachments/files/18904929/coverage-statistics-test.zip)

> On a related note, I am also considering whether we should have some kind of unit-level test for generation of the summary table. When regenerating the expected reports for the end-to-end tests, a possible failure mode is to scan through the many updates too quickly and not notice if a regenerated report is incorrect.

I don't think I know enough about the regeneration/testing side to really answer this at the moment. 
Are you thinking about creating a hand crafted `*-coverage.xml` file so that you know what the summary table numbers should be and any deviation is flagged as a failure? 
Are you thinking of a standard XSpec test that is always manually checked to make sure that the numbers are correct? If so, could the attached test be amended to become that test?

birdya22 reviewed (commented):
Comment added.

galtm commented:
@birdya22 : Thanks for your replies today. It might be a couple of days until I can come back to this pull request.

galtm commented:
> > @birdya22 : Do you have an opinion about whether your TestStylesheet01.xspec file and its related XSLT files should be included in this repo?
> 
> Yes, as I think it is the only test that includes both 'used' and 'not used' modules. It also includes all types of coverage. However, the file names I used aren't useful, so I have amended them and attached the new set.
> 
> [coverage-statistics-test.zip](https://github.com/user-attachments/files/18904929/coverage-statistics-test.zip)

Thanks! 701bac8ca6ef090d335afd8ffcff94765ddeb35f has your files.
 
> > On a related note, I am also considering whether we should have some kind of unit-level test for generation of the summary table. When regenerating the expected reports for the end-to-end tests, a possible failure mode is to scan through the many updates too quickly and not notice if a regenerated report is incorrect.
> 
> I don't think I know enough about the regeneration/testing side to really answer this at the moment. Are you thinking about creating a hand crafted `*-coverage.xml` file so that you know what the summary table numbers should be and any deviation is flagged as a failure? Are you thinking of a standard XSpec test that is always manually checked to make sure that the numbers are correct? If so, could the attached test be amended to become that test?

I was thinking about some kind of hand-crafted data -- yes, so we'd know what the summary tables should be and any deviation becomes a failure. No manual checking needed.

I thought of using as a starting point the `*-coverage.xml` file from your test. I ran into some problems:

1. In `coverage-compute-status.xsl`, the expression `key('modules', $uri, $trace)` relies on an absolute URI, but test files can't use absolute URIs or else what works on my computer won't work in GitHub and vice versa. I addressed that with an alternative `xsl:key` declaration.
2. The line number and column number functions weren't being found (`XPST0017  Cannot find a 1-argument function named Q{http://www.jenitennison.com/xslt/xspec}line-number()` and similar for column-number). Probably that is surmountable (maybe by adjusting the classpath used to run the test), but it made me want to test the table construction code in isolation from the coverage computation code.

So, I ended up with an XSpec test `external_coverage-contents-table.xspec` that uses hand-crafted values for the `coverage-stats` variable and doesn't reach the coverage computation code. I moved some of your code to a separate XSLT template, to provide an entry point for XSpec. In the test scenarios, the numbers of `<coverage>` elements are small, making it easy to see why each node count in the table is what it is. If you have comments, please let me know!

galtm commented:
@birdya22 , I see your thumbs-up on my last comment. Is that equivalent to an "approve" code review? If you were merely acknowledging the comment and still plan to review the files further, I will wait.

birdya22 commented:
That is equivalent to "approve".

galtm commented:
> That is equivalent to "approve".

Great, thank you. I'll merge this today.
